### DISCUSSION TITLE
Need help assigning tags to virtual machines via api

### DISCUSSION THREAD

**Discussion by MrgampRus** [netbox-community/netbox#11588] [Q&A]
When I try to send a patch to a virtual machine with a tag assigned, I get errors, I send the request via powershell. Tell me how to correctly fill in the tags for sending. In the example, I'm trying to assign tag = MES to the virtual machine, the tag exists in the netbox system, I tried to fill the tag with different options, but it doesn't work. Tell me what I'm doing wrong.
![tag1](https://user-images.githubusercontent.com/122436953/214767935-fb23bc23-3a1b-45cb-86fe-a4ffc7d6b70c.JPG)
![tag2](https://user-images.githubusercontent.com/122436953/214767938-e5404f02-96b4-4552-8cc6-934216671f55.JPG)

**Comment by candlerb** [comment]
"tags" needs to be a list, not an object.

Try printing $vCentreJSON before sending it.  To me, it looks like your code has built the following:

```
{"tags": {"name": "MES"}}
```

But it should be:

```
{"tags": ["MES"]}
```

(Aside: I believe this will replace *all* the tags on the object, so if there are other tags, it's your responsibility to merge the list first)

**Reply by MrgampRus** [reply]
Thanks, added vCenterJSON, I saw how it converts and it helped to solve. It is not the tag name that needs to be passed to the virtual machine, it is the ID that needs to be passed. @("id"), Could you describe in more detail what I need to do so that when adding a tag it does not delete others, Thank you.

**Reply by candlerb** [reply]
> It is not the tag name that needs to be passed to the virtual machine, it is the ID that needs to be passed

OK, thanks for the correction. I thought I had used the tag slug successfully in the past, but I could be wrong.

I think you ought to be able to provide a list of objects, e.g.

```
{"tags": [{"name": "MES"}]}
```

(but I've not tested this)

> Could you describe in more detail what I need to do so that when adding a tag it does not delete others

As far as I know, you'll need to read the existing VM's list of tags, modify it, and write it back.  I don't know of any "add tag" or "remove tag" operation in the REST API - even though these are available as bulk operations in the web interface. (In general, the web interface doesn't eat the API dogfood; it has its own endpoints)
